pkg/services/grafana-apiserver/auth/authorizer/provider.go

```diff
@@ -1,34 +1,65 @@
 package authorizer
 
 import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apiserver/pkg/authorization/authorizer"
 	"k8s.io/apiserver/pkg/authorization/union"
 
-	"github.com/grafana/grafana/pkg/services/grafana-apiserver/auth/authorizer/impersonation"
-	"github.com/grafana/grafana/pkg/services/grafana-apiserver/auth/authorizer/org"
-	"github.com/grafana/grafana/pkg/services/grafana-apiserver/auth/authorizer/stack"
+	orgsvc "github.com/grafana/grafana/pkg/services/org"
 	"github.com/grafana/grafana/pkg/setting"
 )
 
-func ProvideAuthorizer(
-	orgIDAuthorizer *org.OrgIDAuthorizer,
-	orgRoleAuthorizer *org.OrgRoleAuthorizer,
-	stackIDAuthorizer *stack.StackIDAuthorizer,
-	cfg *setting.Cfg,
-) authorizer.Authorizer {
+var _ authorizer.Authorizer = (*GrafanaAuthorizer)(nil)
+
+type GrafanaAuthorizer struct {
+	apis map[string]authorizer.AuthorizerFunc
+	auth authorizer.Authorizer
+}
+
+func NewGrafanaAuthorizer(cfg *setting.Cfg, orgService orgsvc.Service) *GrafanaAuthorizer {
 	authorizers := []authorizer.Authorizer{
-		&impersonation.ImpersonationAuthorizer{},
+		&impersonationAuthorizer{},
 	}
 
 	// In Hosted grafana, the StackID replaces the orgID as a valid namespace
 	if cfg.StackID != "" {
-		authorizers = append(authorizers, stackIDAuthorizer)
+		authorizers = append(authorizers, newStackIDAuthorizer(cfg))
 	} else {
-		authorizers = append(authorizers, orgIDAuthorizer)
+		authorizers = append(authorizers, newOrgIDAuthorizer(orgService))
 	}
 
+	// Individual services may have explicit implementations
+	apis := make(map[string]authorizer.AuthorizerFunc)
+	authorizers = append(authorizers, &authorizerForAPI{apis})
+
 	// org role is last -- and will return allow for verbs that match expectations
-	// Ideally FGAC happens earlier and returns an explicit answer
-	authorizers = append(authorizers, orgRoleAuthorizer)
-	return union.New(authorizers...)
+	// The apiVersion flavors will run first and can return early when FGAC has appropriate rules
+	authorizers = append(authorizers, newOrgRoleAuthorizer(orgService))
+	return &GrafanaAuthorizer{
+		apis: apis,
+		auth: union.New(authorizers...),
+	}
+}
+
+func (a *GrafanaAuthorizer) Register(gv schema.GroupVersion, fn authorizer.AuthorizerFunc) {
```

Comment:
not sure if this flexibility is needed, but a possible alternative approach might be to make `Register` more generic and wrap each API authorizer with a modified version of `authorizerForAPI` before registering:
```suggestion
func (a *GrafanaAuthorizer) Register(authz authorizer.Authorizer) {
```